subsys/net/ip/connection.c

```diff
@@ -601,8 +601,12 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 			continue;
 		}
 
-		if (IS_ENABLED(CONFIG_NET_UDP) ||
-		    IS_ENABLED(CONFIG_NET_TCP)) {
+		if ((IS_ENABLED(CONFIG_NET_UDP) ||
+		     IS_ENABLED(CONFIG_NET_TCP))
+#if defined(CONFIG_NET_DSA)
+		    && !net_dsa_recv(pkt_iface, pkt)
```

Comment:
This call looks indeed in wrong place. It was a long time I last time reviewed this so have forgotten why do we actually need to have special handling for raw/packet sockets.

Reply:
**So the use case is as follows:**
We do want to have enabled in the same time support for TCP/IP processing (to e.g. have working `icmp` protocol)  and be able to handle some LLDP packets. The former shall be handled by Zephyr's network stack and the latter (LLDP) by threads spawn for each LAN{123} port (as it is presented in the dsa_lldp.c example).

For LLDP we use `socket(AF_PACKET, SOCK_RAW, ETH_P_ALL);` to create socket.
The code snippet above checks if received packet's MAC address matches the one which was written to KSZ8794's static table `01:80:c2:00:00:03`. The check is done in `dsa_ll_addr_switch_cb` (./samples/net/dsa/src/main.c) callback. If MAC
address marches - then the packet is NOT passed to Zephyr's network stack and is passed directly to LAN{123} interface (L687 in the same file) . 

This is "a hack", as the network stack shall assess if `AF_PACKET && SOCK_RAW` are used with socket creation and 
by itself pass packet to LAN{123} (this is at least how it works in Linux, from which the use case code was ported).
In the presented solution I do reuse the code, which sets the KSZ8794 static table and workaround lack of `AF_PACKET && SOCK_RAW` support with checking MAC address of incoming packets.

On the contrary, when I send the packet from lanX interfaces (`iface`) in the example code, the entry in static table forces using switch tag to forward the packet (with proper MAC) to the specified port. In that way I do achieve the most wanted feature - port separation between LAN{123} ports in the switch.

Reply:
If you only want to catch LLDP packets, we already have functions for that. See `net_lldp_register_callback()` and friends in `include/net/lldp.h`

Reply:
LLDP is only the example - for which I was able to write a decent sample code.
In my use case - the work is performed on customized frames. In the past I've considered the code from ./samples/net/lldp/src/, but it did not fit.

Reply:
Could we discuss how this problem (i.e. lack of `AF_PACKET && SOCK_RAW`) could be solved? Any roadmap or possible obstacles could be helpful.

Reply:
> Could we discuss how this problem (i.e. lack of `AF_PACKET && SOCK_RAW`) could be solved?

Yes, agreed. I have not forgotten this one :)

Naturally we need to support the combination of packet sockets and normal inet/inet6 sockets at the same time. After looking the code I think we could do this something like this:

* Modify/remove `packet_socket.c` so that packet is feed to connection.c without modifications. It seems that we might not need this file at all if the connection.c supports raw sockets properly.
* In `net_conn_input()`, when checking the recipient socket for the packet, check if there are any raw socket listeners and clone the packet (so that it can be delivered to multiple recipients). Similar thing is already done for multicast sockets (line 660), and for dgram type packet sockets (line 693).

The changes might actually be not that big. Would you be interest in implementing this?

Reply:
I will look on this closer after I restructure the code for DSA (i.e. put callbacks into proper structures).